File: LeetCode/p491.cpp
class Solution {
public:
    vector<vector<int>> findSubsequences(vector<int> &nums) {
        raw_data = nums;
        len = nums.size();
        used = vector<bool>(len);
        backtracking(0);
        return result;
    }

private:
    int len;
    vector<int> raw_data;
    vector<bool> used;
    vector<int> tmp_vec;
    vector<vector<int>> result;

    void backtracking(int start) {
        if (tmp_vec.size() >= 2)
            result.emplace_back(tmp_vec);
        if (start == len)
            return;
        unordered_set<int> appeared;
        for (int i = start; i < len; ++i) {
            if (appeared.count(raw_data[i]))
                continue;
            if (!tmp_vec.empty() && raw_data[i] < tmp_vec.back())
                continue;
            appeared.insert(raw_data[i]);
            tmp_vec.push_back(raw_data[i]);
            used[i] = true;
            backtracking(i + 1);
            used[i] = false;
            tmp_vec.pop_back();
        }
    }
};

File: LeetCode/p491.rs
use std::collections::HashSet;

struct Combine {
  nums: Vec<i32>,
  tmp: Vec<i32>,
  ret: Vec<Vec<i32>>,
}

impl Combine {
  pub fn combine(mut self) -> Vec<Vec<i32>> {
    self.backtrack(0);
    self.ret
  }

  fn backtrack(&mut self, start: usize) {
    if self.tmp.len() >= 2 {
      self.ret.push(self.tmp.clone());
    }
    if start == self.nums.len() {
      return;
    }
    let mut used = HashSet::<i32>::new();
    for i in start..self.nums.len() {
      if !self.tmp.is_empty() && self.tmp.last().unwrap() > &self.nums[i] || used.contains(&self.nums[i]) {
        continue;
      }
      self.tmp.push(self.nums[i]);
      used.insert(self.nums[i]);
      self.backtrack(i + 1);
      self.tmp.pop();
    }
  }
}

impl Solution {
  pub fn find_subsequences(nums: Vec<i32>) -> Vec<Vec<i32>> {
    Combine { nums, tmp: vec![], ret: vec![] }.combine()
  }
}

